### ISSUE TITLE
Allow Keras Callbacks to access predictions on_batch_end, on_epoch_end

### ISSUE THREAD

**Issue by evictor** [tensorflow/tensorflow#21174]
### System information

- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: n/a
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.9.0-rc0-35-g17d6639b55 1.9.0-rc1
- **Python version**: 3.5.2
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**: n/a

### Describe the problem

I wanted to call attention to an issue and a PR on the keras-team/keras project so that TF could track and integrate with TF-core Keras. I believe these issues are of high importance to the community; for my project at least they are just short of critical. In summary:

* PR: Keras Callbacks should be able to access predictions made during training in `on_batch_end()` and `on_epoch_end()` to eliminate expensive duplicate calls to `predict()` in Callbacks that require prediction results -- https://github.com/keras-team/keras/pull/10513
* Writing a Keras metric should not require a graph -- https://github.com/keras-team/keras/issues/4506#issuecomment-405452870

The reason these are related is because a viable solution to the second issue depends on the first; in particular, if you wanted to write a "fancy" metric function that did calcs in Python, for example, you could leverage already-made predictions from the API described in issue 1 and circumvent having to `predict()` all over again, which is prohibitively expensive in all but the simplest models.

**Comment by cy89** [comment]
@fchollet PTAL.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by doiko** [comment]
Accessing true and predict data on_epoch_end callback will probably not  be possible (tf is not storing them and size will be number_batches*batch_size*size_sample).
Easy way to access the tf predicted and targets stored in model_targets and model.outputs should be considered. See also  #32981

**Comment by evictor** [comment]
@doiko Preds are at least already available without extra overhead for on_batch_end, see linked line: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training_arrays.py#L304

When I get a free few minutes I will submit a PR exposing these to on_batch_end; it is a trivial change now that I am looking at the code.

I can see what you mean re: not exposing to on_epoch_end, which would require accumulation of outputs. Perhaps that could be an opt-in flag in Callback. I know at least for my use case, though predict itself requires tons of resources, the predictions themselves are small and an epoch's worth will make negligible impact on memory.

**Comment by RoelantStegmann** [comment]
Would be a cool function - also her predictions are a small overhead compared to the overhead of prediction time.

**Comment by Permafacture** [comment]
Just having the predictions and truth in the callback's on_batch_end as @evictor suggests would be a great start.  The callback can accumulate them if desired, and it wouldn't be hard for users to put this logic into a parent MetricCallback class if they desire.  The current solution of instantiating the callback with the training data and truth is much more awkward.

Also, @evictor you linked to a line number in Master, and since master has been updated since your post, I don't think that line is what you meant anymore.

**Comment by rmothukuru** [comment]
@evictor,
Can you please confirm if [on_predict_batch_begin](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#on_predict_batch_begin)  and [on_predict_batch_end](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#on_predict_batch_end) are what you are looking for?

Thanks!

**Comment by evictor** [comment]
It's been a while since I worked on that project, but yes this appears to solve the original problem we faced.

**Comment by rmothukuru** [comment]
@evictor,
Thank you for your confirmation. Closing the issue.
